### ISSUE TITLE
docker builder fails to copy from `uploadData` when running in nested container. hangs forever

### ISSUE THREAD

**Issue by emalloy** [mitchellh/packer#3102]
Had this problem with 0.8.6-release, so did researching and building v0.8.7.dev (b7ba9a5) and getting same issue -

when building from within a docker container...  _who is connecting to the parent docker daemon via exposed unix socket

os is debian jessie, building an ubuntu foundation image from ubuntu:latest

problem is with inline and shell provisioner. haven't got as far as ansible_local provisioner yet, but i imagine that is the same issue.....

the invocation of the container at runtime for which packer is running under looks like this:

```
docker run -P -d --privileged -it  -v /lib/x86_64-linux-gnu/libdevmapper.so.1.02.1:/lib/x86_64-linux-gnu/libdevmapper.so.1.02.1 -v /var/run/docker.sock:/var/run/docker.sock -v /usr/bin/docker:/usr/bin/docker docker2.moammo.net:5000/jenkins-exp:0.2 
```

the official jenkinsci docker image is a parent of docker2.moammo.net:5000/jenkins-exp:0.2

from the top of my packer builder.json

```
{
  "variables": {
    "imageTag": "",
    "imageRepo": ""
  },
  "builders": [
    {
      "type": "docker",
      "image": "ubuntu:latest",
      "commit": true,
      "run_command": [
        "-ti",
        "-v",
        "/tmp/packerfiles:/tmp/packerfiles",
        "-d",
        "{{.Image}}",
        "/bin/bash",
        "--noprofile",
        "--norc"
      ]
    }
  ],
  "provisioners": [
    {
      "type": "shell",
      "scripts": [
        "packer/scripts/docker_ubuntu.sh"
      ]
    },
```

my packer invocation looks like this:

```
VERSION=$1
IMGREPO=$2
JSONFILE=$3
mkdir -p /tmp/packerfiles
TMPDIR=/tmp/packerfiles PACKER_LOG=1 packer build \
  -var imageTag=${VERSION} \
  -var imageRepo=${IMGREPO} \
  packer/scripts/${JSONFILE}.json
exit $?
```

output from PACKER_LOG=1

```
2016/01/22 18:45:44 ui: �[1;32m==> docker: Starting docker container...�[0m
2016/01/22 18:45:44 ui: �[0;32m    docker: Run command: docker run -v /var/jenkins_home/.packer.d/tmp/packer-docker523294758:/packer-files -ti -v /tmp/packerfiles:/tmp/packerfiles -d ubuntu:latest /bin/bash --noprofile --norc�[0m
�[0;32m    docker: Status: Image is up to date for ubuntu:latest�[0m
�[1;32m==> docker: Starting docker container...�[0m
2016/01/22 18:45:44 packer: 2016/01/22 18:45:44 Starting container with args: [run -v /var/jenkins_home/.packer.d/tmp/packer-docker523294758:/packer-files -ti -v /tmp/packerfiles:/tmp/packerfiles -d ubuntu:latest /bin/bash --noprofile --norc]
�[0;32m    docker: Run command: docker run -v /var/jenkins_home/.packer.d/tmp/packer-docker523294758:/packer-files -ti -v /tmp/packerfiles:/tmp/packerfiles -d ubuntu:latest /bin/bash --noprofile --norc�[0m
2016/01/22 18:45:44 packer: 2016/01/22 18:45:44 Waiting for container to finish starting
2016/01/22 18:45:47 ui: �[0;32m    docker: Container ID: 6a9448c9580ac223dc4b8e5dbc7e174cc888293bf814e254cafe09c99dfe2570�[0m
�[0;32m    docker: Container ID: 6a9448c9580ac223dc4b8e5dbc7e174cc888293bf814e254cafe09c99dfe2570�[0m
2016/01/22 18:45:47 packer: 2016/01/22 18:45:47 Running the provision hook
2016/01/22 18:45:47 ui: �[1;32m==> docker: Provisioning with shell script: packer/scripts/docker_ubuntu.sh�[0m
2016/01/22 18:45:47 packer: 2016/01/22 18:45:47 Opening packer/scripts/docker_ubuntu.sh for reading
�[1;32m==> docker: Provisioning with shell script: packer/scripts/docker_ubuntu.sh�[0m
2016/01/22 18:45:47 packer: 2016/01/22 18:45:47 [INFO] 1035 bytes written for 'uploadData'
2016/01/22 18:45:47 [INFO] 1035 bytes written for 'uploadData'
2016/01/22 18:45:47 packer: 2016/01/22 18:45:47 Executing in container 6a9448c9580ac223dc4b8e5dbc7e174cc888293bf814e254cafe09c99dfe2570: "(command cp /packer-files/upload507966797 /tmp/script_898.sh) >/packer-files/cmd213455432 2>&1; echo $? >/packer-files/cmd213455432-exit"
2016/01/22 18:45:49 packer: 2016/01/22 18:45:49 Waiting for exit code to appear for remote command...
```

I enter the container and inspect /packer-files

```
root@docker3:~# docker exec -it big_kilby /bin/bash
root@6a9448c9580a:/# pushd /tmp/packerfiles/
/tmp/packerfiles /
root@6a9448c9580a:/tmp/packerfiles# ls -lah
total 16K
drwxr-xr-x 4 root root 4.0K Jan 22 18:35 .
drwxrwxrwt 3 root root 4.0K Jan 22 18:45 ..
drwxr-xr-x 2 root root 4.0K Jan 22 18:35 packer-docker329986877
drwxr-xr-x 2 root root 4.0K Jan 22 16:50 packer-docker463191192
root@6a9448c9580a:/tmp/packerfiles# cat packer-docker329986877/cmd749252215
cmd749252215       cmd749252215-exit
root@6a9448c9580a:/tmp/packerfiles# cat packer-docker329986877/cmd749252215*
cp: cannot stat '/packer-files/upload772259704': No such file or directory
1
root@6a9448c9580a:/tmp/packerfiles# pushd /packer-files/; ls -lah
/packer-files /tmp/packerfiles /
total 16K
drwxr-xr-x  2 root root 4.0K Jan 22 18:45 .
drwxr-xr-x 22 root root 4.0K Jan 22 18:45 ..
-rw-r--r--  1 root root   75 Jan 22 18:45 cmd213455432
-rw-r--r--  1 root root    2 Jan 22 18:45 cmd213455432-exit
root@6a9448c9580a:/packer-files# cat cmd213455432*
cp: cannot stat '/packer-files/upload507966797': No such file or directory
1
```

**Comment by lv0** [comment]
interesting as well ,if i run with `-debug` switch, 

```
2016/01/22 21:06:41 packer: 2016/01/22 21:06:41 [DEBUG] Docker version: 1.9.1
2016/01/22 21:06:41 packer: 2016/01/22 21:06:41 [DEBUG] Container will be committed
==> docker: Creating a temporary directory for sharing data...
2016/01/22 21:06:41 ui: ==> docker: Creating a temporary directory for sharing data...
2016/01/22 21:06:41 ui: ask: ==> docker: Pausing after run of step 'StepTempDir'. Press enter to continue.
2016/01/22 21:06:41 packer: 2016/01/22 21:06:41 Detected home directory from env var: /var/jenkins_home
==> docker: Pausing after run of step 'StepTempDir'. Press enter to continue.


2016/01/22 21:07:31 ui: scan err: unexpected newline
2016/01/22 21:07:31 ui: ==> docker: Pulling Docker image: ubuntu:latest
2016/01/22 21:07:31 packer: 2016/01/22 21:07:31 Executing: /usr/bin/docker [pull ubuntu:latest]
```

which after some delay reports:

```
2016/01/22 21:07:31 ui: scan err: unexpected newline
```

which i seen in a closed issue about an scp problem. we should be using `docker cp <file> <container>:<file>` no? not scp?

**Comment by cbednarski** [comment]
> which i seen in a closed issue about an scp problem. we should be using docker cp <file> <container>:<file> no? not scp?

No, we've actually implemented this differently because, IIRC, `docker cp` was not supported when this feature was implemented. From memory (so this may be wrong) uploads use SCP and downloads use `docker cp` (download was implemented later). If we switch to the native docker way to do it we have to also increase the minimum docker version that packer will work with to a version that supports `docker cp`.

This is something we'll likely look at later but I think this code path needs to be rewritten (perhaps on both sides) to use [go-dockerclient](https://github.com/fsouza/go-dockerclient).

**Comment by lv0** [comment]
Thanks for responding @cbednarski 

To use packer to deliver docker foundation images in a CI pipeline, this may be a necessary feature.

With a lot of the CI work being done inside containers, the `docker cp` delivery over the docker exposed/bound socket may be the only way to get this done, for this ever-popular use-case.

Do you have any suggestions for the interim, in this situation?

TIA!

**Comment by cbednarski** [comment]
> Do you have any suggestions for the interim, in this situation?

Not off-hand. Changing the implementation to use the docker API over the socket is the most straightforward fix.

I suspect the reason this isn't working is because packer expects to be able to directly communicate with the docker container that it is building, but the network looks different from inside the container as compared to running on the host. Essentially, packer can see the docker socket but not the container it's building, so it hangs.

You might be able to get this to work by tweaking your docker network options so the container running packer uses the same network as the host (e.g. _not_ `docker0`) and sees the running container the from the host's perspective.

The workaround I would recommend for now is not running packer in a container. I understand the use-case you have and while we can eventually improve this _specific_ use-case, running packer-in-docker is not likely to ever be supported for some cases like VMware, VirtualBox, etc. which require some level of root access on the host (also because of networking, among other reasons).

**Comment by dgalichet** [comment]
I've got the same kind of issue when trying to use a `file provisioner` to copy a local file to a container.
I'm running `docker-machine` under OSX and the `packer build` stuck indefinitely during copy of a local file to the container. `docker cp` looks like a promising solution

**Comment by tylert** [comment]
This might also be related to #2497.

**Comment by jfusterm** [comment]
I've faced the same issue with both `file` and `shell` provisioners. I'm running a Docker image with Packer installed, that has all the plugins that I need to build other Docker images in a typical DooD (*Docker-outside-of-Docker*) scenario.

```
$ packer -v
0.11.0

$ docker --version
Docker version 1.12.2, build bb80604
```

Starting the Packer container.

```
$ docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v /packer-templates:/tmp my-image-with-packer
```

Once I'm in the container, I can create and use containers or build images on the Docker host. I have no problems whatsoever with either using or not `--privileged` or bind-mounting the Docker binary with `-v $(which docker):/usr/local/bin/docker`. It just works.

But when I'm trying to use Packer within the container to build another Docker images on the Docker host, it fails on the `file` and `shell` provisioner.


Using the `shell` provisioner.

```
{
    "builders": [
        {
            "type": "docker",
            "image": "debian:jessie",
            "commit": true
        }
    ],

    "provisioners": [
        {
            "type": "shell",
            "only": ["docker"],
            "inline": [
                "set -ex",
                "date > date.txt"
            ]
        }
    ],

    "post-processors": [
        [
            {
                "type": "docker-tag",
                "only": ["docker"],
                "repository": "test/debian",
                "tag": "packer"
            }
        ]
    ]
}
```



```
$ PACKER_LOG=1 packer build template.json

...
2016/10/26 10:13:10 ui:     docker: Container ID: 55b1ec606f9e864be392ed9f3088031a87af1c4cecf03117eed7d484a7e22f09
    docker: Container ID: 55b1ec606f9e864be392ed9f3088031a87af1c4cecf03117eed7d484a7e22f09
2016/10/26 10:13:10 packer: 2016/10/26 10:13:10 Running the provision hook
==> docker: Provisioning with shell script: /tmp/packer-shell430759610
2016/10/26 10:13:10 ui: ==> docker: Provisioning with shell script: /tmp/packer-shell430759610
2016/10/26 10:13:10 packer: 2016/10/26 10:13:10 Opening /tmp/packer-shell430759610 for reading
2016/10/26 10:13:10 packer: 2016/10/26 10:13:10 [INFO] 33 bytes written for 'uploadData'
2016/10/26 10:13:10 [INFO] 33 bytes written for 'uploadData'
2016/10/26 10:13:10 packer: 2016/10/26 10:13:10 Executing docker exec -i 55b1ec606f9e864be392ed9f3088031a87af1c4cecf03117eed7d484a7e22f09 /bin/sh -c (command cp /packer-files/upload253254180 /tmp/script_4928.sh):
2016/10/26 10:13:10 packer: 2016/10/26 10:13:10 Retryable error: Error uploading script: Upload failed with non-zero exit status: 1
2016/10/26 10:13:12 packer: 2016/10/26 10:13:12 [INFO] 33 bytes written for 'uploadData'
2016/10/26 10:13:12 [INFO] 33 bytes written for 'uploadData'
2016/10/26 10:13:12 packer: 2016/10/26 10:13:12 Executing docker exec -i 55b1ec606f9e864be392ed9f3088031a87af1c4cecf03117eed7d484a7e22f09 /bin/sh -c (command cp /packer-files/upload930801459 /tmp/script_4928.sh):
2016/10/26 10:13:12 packer: 2016/10/26 10:13:12 Retryable error: Error uploading script: Upload failed with non-zero exit status: 1
2016/10/26 10:13:14 packer: 2016/10/26 10:13:14 [INFO] 33 bytes written for 'uploadData'
2016/10/26 10:13:14 [INFO] 33 bytes written for 'uploadData'
2016/10/26 10:13:14 packer: 2016/10/26 10:13:14 Executing docker exec -i 55b1ec606f9e864be392ed9f3088031a87af1c4cecf03117eed7d484a7e22f09 /bin/sh -c (command cp /packer-files/upload579890934 /tmp/script_4928.sh):
2016/10/26 10:13:14 packer: 2016/10/26 10:13:14 Retryable error: Error uploading script: Upload failed with non-zero exit status:
...
```

And it gets stuck forever.

Using the `file` provisioner.

```
{
    "builders": [
        {
            "type": "docker",
            "image": "debian:jessie",
            "commit": true
        }
    ],

    "provisioners": [
        {
            "type": "file",
            "only": ["docker"],
            "source": "scripts/",
            "destination": "/scripts/"
        }
    ],

    "post-processors": [
        [
            {
                "type": "docker-tag",
                "only": ["docker"],
                "repository": "test/debian",
                "tag": "packer"
            }
        ]
    ]
}
```

```
$ PACKER_LOG=1 packer build template.json
...
2016/10/26 10:18:10 ui: ==> docker: Starting docker container...
==> docker: Starting docker container...
2016/10/26 10:18:10 ui:     docker: Run command: docker run -v /root/.packer.d/tmp/packer-docker224977642:/packer-files -d -i -t debian:jessie /bin/bash
2016/10/26 10:18:10 packer: 2016/10/26 10:18:10 Starting container with args: [run -v /root/.packer.d/tmp/packer-docker224977642:/packer-files -d -i -t debian:jessie /bin/bash]
2016/10/26 10:18:10 packer: 2016/10/26 10:18:10 Waiting for container to finish starting
    docker: Run command: docker run -v /root/.packer.d/tmp/packer-docker224977642:/packer-files -d -i -t debian:jessie /bin/bash
2016/10/26 10:18:11 ui:     docker: Container ID: 8457f7ba27e71d484a54ee5fad004b3e4da42f036e17683bd370c6d45db3124a
    docker: Container ID: 8457f7ba27e71d484a54ee5fad004b3e4da42f036e17683bd370c6d45db3124a
2016/10/26 10:18:11 packer: 2016/10/26 10:18:11 Running the provision hook
==> docker: Uploading scripts/ => /scripts/
2016/10/26 10:18:11 ui: ==> docker: Uploading scripts/ => /scripts/
2016/10/26 10:18:11 packer: 2016/10/26 10:18:11 Executing docker exec -i 8457f7ba27e71d484a54ee5fad004b3e4da42f036e17683bd370c6d45db3124a /bin/sh -c (set -e; mkdir -p /scripts/; cd /packer-files/dirupload638483265; command cp -R `ls -A .` /scripts/):
2016/10/26 10:18:11 ui: ==> docker: Killing the container: 8457f7ba27e71d484a54ee5fad004b3e4da42f036e17683bd370c6d45db3124a
==> docker: Killing the container: 8457f7ba27e71d484a54ee5fad004b3e4da42f036e17683bd370c6d45db3124a
2016/10/26 10:18:11 ui error: Build 'docker' errored: Upload failed with non-zero exit status: 2
2016/10/26 10:18:11 Builds completed. Waiting on interrupt barrier...
2016/10/26 10:18:11 machine readable: error-count []string{"1"}
2016/10/26 10:18:11 ui error:
==> Some builds didn't complete successfully and had errors:
2016/10/26 10:18:11 machine readable: docker,error []string{"Upload failed with non-zero exit status: 2"}
2016/10/26 10:18:11 ui error: --> docker: Upload failed with non-zero exit status: 2
2016/10/26 10:18:11 ui:
==> Builds finished but no artifacts were created.
2016/10/26 10:18:11 waiting for all plugin processes to complete...
Build 'docker' errored: Upload failed with non-zero exit status: 2
2016/10/26 10:18:11 /usr/local/bin/packer: plugin process exited
2016/10/26 10:18:11 /usr/local/bin/packer: plugin process exited
2016/10/26 10:18:11 /usr/local/bin/packer: plugin process exited

==> Some builds didn't complete successfully and had errors:
--> docker: Upload failed with non-zero exit status: 2

==> Builds finished but no artifacts were created.
```

The build fails with exit `2`.

**Comment by mwhooker** [comment]
I think this will be resolved by #4186. After that gets merged, please open a new issue if you continue to see this
